service/operator/service.go

```diff
@@ -157,10 +169,37 @@ func (s *Service) addFunc(obj interface{}) {
 }
 
 func (s *Service) deleteFunc(obj interface{}) {
+	// We lock the addFunc/deleteFunc/updateFunc to make sure only one
+	// addFunc/deleteFunc/updateFunc is executed at a time.
+	// addFunc/deleteFunc/updateFunc is not thread safe. This is important because
+	// the source of truth for the azure-operator are Azure resources.
+	// In case we would run the operator logic in parallel, we would run into race
+	// conditions.
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
 	s.logger.Log("debug", "executing the operator's deleteFunc")
 
 	err := s.operatorFramework.ProcessDelete(obj, s.resources)
 	if err != nil {
 		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "delete")
 	}
 }
+
+func (s *Service) updateFunc(oldObj interface{}, newObj interface{}) {
```

Comment:
I think body of this function should be `addFunc(newObj)`. I strongly disagree we should handle updates separately. It's unreliable, and should be used only as an optimization. I'd even remove `ProcessUpdate` from the framework as I said in another review.

If we call `addFunc(newObj)` here we'll force ourselves to create an idempotent operator from the very begging which is a very good thing.

Reply:
I get the issue with `ProcessUpdate` and I think we should change the updateFunc to call `ProcessCreate`. At the moment updates are not required we just need to be able to create / delete resource groups and deployments.

We need to establish consensus on how we handle updates and if we're not going to use it we should remove `ProcessUpdate` from the framework.

From my POV I think there is benefit from detecting if a deployment has changed and only then launching a new deployment. But maybe we can just always launch a new deployment by calling `CreateOrUpdate` in the Azure SDK.

Reply:
The updateFunc is updated.

Reply:
> From my POV I think there is benefit from detecting if a deployment has changed and only then launching a new deployment.

I think this detection should be the operator responsibility. We have no guarantee that we receive every object update. And in many cases object will arrive to the `addFunc` despite fact it already exists, i.e. after `ResyncPeriod` and when the operator is restarted.

Reply:
Yes, I think that can and should be the operator's responsibility. I think your point on the `addFunc` being called out of sequence is key. Let's continue this as part of the `operatorkit` discussions.

Reply:
Moreover updating functionality comes from the `cache` (client) package. This is not K8s API feature.